pkg/cover/report_test.go

```diff
@@ -128,6 +128,9 @@ func TestReportGenerator(t *testing.T) {
 func testReportGenerator(t *testing.T, target *targets.Target, test Test) {
 	rep, csv, err := generateReport(t, target, test)
 	if err != nil {
+		if target.Arch == targets.ARM64 && err.Error() == "coverage doesn't match any coverage callbacks" {
```

Comment:
This is not good. We are losing test coverage, so more things become broken. And any such disabling needs an explanatory comment explaining why it's disabled.
What's wrong with the test? Is it possible to fix the test instead of disabling it?

Reply:
So fixupPcs for arm64 will change the pc and so err will be "coverage doesn't match any coverage callbacks".
Other possible fix?

Reply:
This error means that no coverage report will be generated for arm64.
The solution depends on _why_ pcs don't match callbacks now. E.g. if fixupPcs produces bogus PCs, then we should remove fixupPcs. One way or another PCs must match coverage callbacks, otherwise no report will be generated.
What's the underlying issue here? Are PCs wrong? Are coverage callbacks wrong?

Reply:
The issue here is the fixupPCs ```
				if pc < 0xffffffd010000000 {
					pc -= 0x18
				}
```

so when the test runs for arm64, as the pc is less than 0xffffffd010000000 , so final pc will be excluded as progPCs[pc] will always be false, and so !matchedPC will be true.

Reply:
I had memory that coverage reports were working on arm64 when I tested (unfortunately we don't have KCOV enabled on syzbot arm64). If they were indeed working, does it mean this change will break them?
What is 0xffffffd010000000? I don't see it in https://www.kernel.org/doc/html/latest/arm64/memory.html

Reply:
The VA_BITS is 39 in my case, so it's 0xffffff1000000000, not sure where is remaining bits comes from.

Reply:
What is 0xffffff1000000000? And what is 0xffffffd010000000?

Reply:
Sorry, it's 0xffffff8000000000 for 39 bits VA_BITS. so for kernel address, if VA_BITS is 39, then above (64-39) bits are set to 1, then it's 0xffffff8000000000 (MSB 25bits set to 1).

And on real device for arm64, I got 
```
$ nm tree1/vmlinux|grep -w _text
ffffffc010000000 T _text
$ nm tree2/vmlinux | grep -w _text
ffffffd010080000 T _text
```

I think it's related to the difference in two arch/arm64/kernel/vmlinux.lds.S files.
So previous fixupPCs is to -0x18 for modules on ARM64 (not found a good way to decide if the pc belongs to a module in that PR. So I trade the PC belongs to module if PC less than _text address), and brining in that one makes the test fail now as that test has PC value about 60000.

Reply:
To avoid hardcode, I think we can pass rg to fixupPC, and the rg can keep the _text address during elf parsing?

Reply:
> To avoid hardcode, I think we can pass rg to fixupPC, and the rg can keep the _text address during elf parsing?

This looks good to me.
It would avoid the hardcoded constant, make the code more self-explanatory (if we use .text address, it's clear what we mean) and I think it will also fix the text, right? So we will keep test coverage for at least some parts of arm64 code.

Reply:
uploaded, but I'm facing race condition during test, as one thread write to progs[i] = prog while another read in prepareFileMap. can you suggest how to fix?

Reply:
The safest and simplest solution would be for fixUpPCs to build a modified copy of progs and returns it and don't mutate the passed in progs slice.

Reply:
done